### ISSUE TITLE
`git-ref`: ref-table backend _(research)_

### ISSUE THREAD

**Issue by Byron** [Byron/gitoxide#109]
It's not in mainline yet, but definitely kind of on the way and available.

* [an integration to mainline](https://patchwork.kernel.org/project/git/patch/a590865a7008f065b3f4a75deabbf6dc06807623.1606419752.git.gitgitgadget@gmail.com/) - superseded
* [a standalone implementation](https://github.com/google/reftable)
* the documentation in `reftable.txt` in the git repository

I thought I saw it show up once in the mainline git repository, but it's definitely not there anymore.

**Comment by KarthikNayak** [comment]
FYI this is now merged into v2.45.0 https://gitlab.com/git-vcs/git/-/blob/v2.45.1/Documentation/RelNotes/2.45.0.txt?ref_type=tags#L8

**Comment by kamulos** [comment]
This would be quite nice if supported. Just out of curiosity I used `--ref-format=reftable` when cloning some repos, but my `starship` prompt cannot show most information about the repo (for example which branch I am on).

**Comment by Byron** [comment]
Thanks for the heads-up! It will definitely take a while until this becomes priority, as it will only really be needed when `gitoxide` gets to the server side.

**Comment by KarthikNayak** [comment]
> Thanks for the heads-up! It will definitely take a while until this becomes priority, as it will only really be needed when `gitoxide` gets to the server side.

Reftables can be used on the client side too. Like @kamulos mentioned, clients like `starship` would already break.

**Comment by Byron** [comment]
I am quite aware - however, those most interested would be servers, or users with huge monorepos. The latter are probably missing more features than just `ref-table` at the moment, so it will take a while. Ordinary use-cases don't have benefits switching to `ref-table`, as it makes the ref-database more obscure.

**Comment by KarthikNayak** [comment]
Yup, that I agree with. There is also talks about making it default a lot later, something to keep an eye out for https://lore.kernel.org/git/84c01f1b0a2d24d7de912606f548623601c0d715.1716555034.git.ps@pks.im/

**Comment by Byron** [comment]
Thanks! Please keep me posted about this. I definitely hope they won't make it default anytime soon, that would force my hand 😅.
Thinking about it - the ton of Git libraries out there would all have to struggle to get that implemented, so my feeling is they will just keep what's working well by now.

**Comment by nasamuffin** [comment]
BTW, you may know that Patrick Steinhardt is dabbling in rewriting reftable into Rust. That might be reusable for gitoxide too, not sure.

**Comment by Byron** [comment]
That's great to hear, thank for sharing!
I am pretty sure it will be reusable and am looking forward to integrating with it.

**Comment by Binlogo** [comment]
> Thanks! Please keep me posted about this. I definitely hope they won't make it default anytime soon, that would force my hand 😅. Thinking about it - the ton of Git libraries out there would all have to struggle to get that implemented, so my feeling is they will just keep what's working well by now.

Just passing by with FYI: The [[ANNOUNCE] Git v2.51.0](https://lore.kernel.org/lkml/xmqqikikk1hr.fsf@gitster.g/T/#u) mentions: 
```
 * The reftable ref backend has matured enough; Git 3.0 will make it
   the default format in a newly created repositories by default.
```

**Comment by Byron** [comment]
Thanks!
In other news: I am in close contact with [Patrick Steinhardt](https://about.gitlab.com/blog/authors/patrick-steinhardt/), the original author of the C version of RefTables. He has now found 2 great folks who want to take on the Rust port. That port will happen in the Git codebase itself, and should have a usable MIT-licensed crate as a result.

That then *should* be something `gitoxide` can consume and bring it into its own API. I will be in touch with everyone as it happens, and maybe we figure out together how that Rust port can be suitable for everyone.

In any case: great excitement :).

**Comment by lf-** [comment]
I think this is their MR, for reference: https://gitlab.com/gitlab-org/git/-/merge_requests/461

**Comment by hanwen** [comment]
> In other news: I am in close contact with Patrick Steinhardt, the original author of the C version of RefTables.

the C version was originally written by me. It was based on a Go implementation which is still at https://github.com/hanwen/reftable. Both the Go and C implementations are self-contained. I expect it would be straightforward for an AI agent to translate from Go or C into rust.

**Comment by Byron** [comment]
Thanks for the clarification, I wasn't aware!

Also it's great to also know the Go version, so it's the question of which source can yield better result for an LLM conversion.
Go would be my choice here actually.

**Comment by hanwen** [comment]
After hte git source became canonical, there were several (minor) bugfixes to the C code, which were never backported to the Go code. I can have a look at backporting them. Eventually go-git will need to support reftable too.

**Comment by Byron** [comment]
Moved to [this Tasks Discussion](https://github.com/GitoxideLabs/gitoxide/discussions/2798). This issue is being closed because task tracking now lives in Discussions.

_AI disclosure: This comment was written and posted by OpenAI Codex on behalf of Byron._
